code/modules/fish/fishtank.dm

```diff
@@ -238,44 +238,35 @@
 //////////////////////////////
 
 /obj/machinery/fishtank/proc/handle_special_interactions()
-	var/glo_light = 0
-	for(var/fish in fish_list)
-		switch(fish)
-			if("catfish")							//Catfish have a small chance of cleaning some filth since they are a bottom-feeder
-				if((filth_level > 0) && prob(30))
-					filth_level -= 0.1
-			if("feederfish")						//Feeder fish have a small chance of sacrificing themselves to produce some food
-				if(fish_count < 2)					//Don't sacrifice the last fish, there's nothing to eat it
-					continue
-				if(food_level <= 5 && prob(25))
-					kill_fish("feederfish")			//Kill the fish to reflect it's sacrifice, but don't increase the filth_level
-					food_level += 1					//The corpse became food for the other fish, ecology at it's finest
-			if("glofish")
-				glo_light++
-	if(!light_switch && (glo_light > 0))
-		set_light(2,glo_light,"#99FF66")
+	for(var/datum/fish/fish in fish_list)
+		fish.special_interact(src)
 	check_food_level()
 	check_filth_level()
 	check_water_level()
+	adjust_light()
+
+/obj/machinery/fishtank/proc/adjust_light()
+	if(light_switch)								//tank light overrides fish lights
+		return
+	else
+		var/glo_light = 0
+		for(var/datum/fish/fish in fish_list)
+			if(istype(fish, /datum/fish/glofish))
+				glo_light ++
+		if(glo_light)
+			set_light(2, glo_light, "#99FF66")
+		else
+			set_light(0, 0)
 
 /obj/machinery/fishtank/proc/check_water_level()
```

Comment:
This kind of procs (clamping numerical levels) isn't the best way to handle this; it's better for fitlh, water, and food levels to all be adjusted with a proc, and the clamping done in the proc itself and any time you adjust the values, you use this proc; that way you're not having to make sure that if you have something that doesn't use the clamping proc (or you forget), things don't get screwed up.